tests/mock_crust/churn.rs

```diff
@@ -90,24 +85,61 @@ fn add_nodes_and_poll<R: Rng>(
             pfx
         })
         .collect();
+
+    let mut added_nodes = Vec::new();
     while !prefixes.is_empty() {
+        let proxy_index = if nodes.len() > nodes[0].chain().min_sec_size() {
+            gen_range(rng, 0, nodes.len())
+        } else {
+            0
+        };
+        let bootstrap_config =
+            BootstrapConfig::with_contacts(&[nodes[proxy_index].handle.endpoint()]);
         let node = TestNode::builder(&network)
             .bootstrap_config(bootstrap_config.clone())
             .create();
         if let Some(&pfx) = prefixes.iter().find(|pfx| pfx.matches(&node.name())) {
             assert!(prefixes.remove(&pfx));
-            if allow_add_failure && !rng.gen_weighted_bool(prefixes.len() as u32) {
+            if skip_some_prefixes && !rng.gen_weighted_bool(prefixes.len() as u32) {
                 continue;
             }
-            nodes.push(node);
+            added_nodes.push(node);
         }
     }
 
-    poll_and_resend(&mut nodes, &mut []);
+    added_nodes.drain(..).for_each(|added_node| {
```

Comment:
Although this does what it's supposed to do, it feels convoluted. Why not just:
```rust
for added_node in added_nodes {
    ...
}
```
?